### ISSUE TITLE
Feature: Map constructor parameters

### ISSUE THREAD

**Issue by JeremyCaney** [Ignia/Topics-Library#35]
Currently, the `TopicMappingService` exclusively maps properties, and requires an empty constructor. It would add flexibility, and especially for records, if constructor parameters could also be mapped. 

## Challenges
There are two primary challenges with this. 

### Double Mapping
If a property is mapped via the constructor, how do we prevent it from being double mapped? This is especially true with records, where the constructor parameter might be used to define a property. 

### Order of Operations 
Currently, the object is initialized, immediately added to cache, and then mapped. This would require assembling a collection of mapped values first and assigning them in unison, instead of assigning them as they're evaluated, and only then adding the object to the cache.

### Circular References 
Extrapolated out, the above means that topic associations might be evaluated prior to the root object(s) being created. This will introduce problems with circular references since the original topic can't be mapped until (constructor) dependencies are mapped. 

### Double Mapping
- Ignore mapping properties if they aren’t an `IList` and aren’t set to their default value—_except_ that would likely introduce problems with properties initialized to a specific value. 
- Require `[DisableMapping]` on the property itself to manually prevent this situation.
- Track mapped constructor parameters in a configuration object passed down to `MapProperties()`. This could even be on the `TopicMappingCacheEntry`. That would depend on the parameter names matching their target properties.

### Circular References 
The circular dependency issue could be addressed by limiting what annotations can be applied to parameters. Specifically, this could exclude:
- `[Include()]`, so no associations are mapped,
- `[MapToParent]`, since no parent object will be available.

**Comment by JeremyCaney** [comment]
Topic associations—including topic references and relationship collections—could both be supported so long as they don’t create a circular reference. One way to handle this would be to track the `Topic.Id` and `Type` that are being initialized, and then throw an exception if that `Topic.Id` and `Type` are requested again before the first finishes initialization. 

This could even be done with the existing `MappedTopicCache` by adding e.g an `IsInitializing` property to the `MappedTopicCacheEntry`, and then throwing an exception `TryGetValue()`—and _possibly_ `AddOrGet()`—attempt to retrieve a `MappedTopicCacheEntry` with `IsInitializing` set to `true`. This way, we don’t need to introduce a new way of tracking current mapping state. 

> _Note:_ This would require careful coordination from callers to avoid inadvertently tripping this exception, but as an `internal` class exclusively called by `TopicMappingService`, that should be easy to manage.

**Comment by JeremyCaney** [comment]
Relationships introduce some challenges in terms of initializing the collection. If it’s a concrete type with an empty constructor, that’s no problem. If it’s an `IList<>` or `IEnumerable<>`, we can extract the generic parameters and use them to initialize a `List<>`. The latter is a bit more involved.

**Comment by JeremyCaney** [comment]
Not overriding null values isn’t ideal for avoiding the double-binding issue. Notably, as raised in the initial description, this risks overriding the initial value if it’s not set to `default`. 

A more sophisticated approach is to track the attribute names, and to bypass `SetPropertyValue()` if they were already set in the constructor. This will need to account for subsequent mappings to pick up associations, thus should still make the call if `mapAssociationsOnly` is true. 

This could optionally be stored as a property on `TopicMappingCacheEntry`, as a way of relaying the information. Alternatively, it may be possible for `InitializeObject()` to call `SetProperty()` directly, thus allowing it to pass the list of parameters as part of that call. This will require consideration during implementation.

**Comment by JeremyCaney** [comment]
To support this, the `TopicMappingService` was refactored to use a new `ItemConfiguration` (88e06a73) instead of a `PropertyConfiguration` (88e06a73) so that the helper methods could operate off of either `ParameterInfo` or `PropertyInfo`. Those same methods were updated to use "get" semantics instead of "set" semantics (304525d6), thus separating out how they are set to either `SetPropertyAsync()` (304525d6) or `GetParameterAsync()` (142d765e). With that, the constructor mapping logic was added to the previously introduced `MapAsync(topic, type…)` overload (3b342653), which was first introduced as part of the `[MapAs()]` support (61de3bf4). Finally, the code was updated to make this the single source of view model construction in the `TopicMappingService` (ad34a65b). This logic is validated via a pair of unit tests (8b3f14e6, 4ac35678) using a new `ConstructedTopicViewModel` (3dcbdd62).

As part of this update, I also refactored the `MemberDispatcher` to separate the (de)serialization of attribute values to value types into a new `AttributeValueConverter` library (84a2758e), and updated the `MappedTopicCache` to have a more focused interface (60ee7d72, 5b3a59cb) and to enforce business logic related to instantiating constructors (8f863924, cb41728f, b33b60cc).
